### PR TITLE
chore(common): updates Web & Web-related dependencies

### PR DESCRIPTION
**Pull request by jahorton**
As per https://github.com/keymanapp/keyman/pull/8149#pullrequestreview-1286404446, this PR updates our in-repo TS dependency to 4.9.5, the most recent 'stable' release build.  (5.0 exists, but is still in beta.)

I also ran an `npm audit fix` round; there were two warnings that'd been popping up.

Furthermore, I tweaked the package.json for `resources/build/version` because of a common warning that'd been popping up:

```
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: undefined,
npm WARN EBADENGINE   required: { node: '16' },
npm WARN EBADENGINE   current: { node: 'v17.9.0', npm: '8.5.5' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: undefined,
npm WARN EBADENGINE   required: { node: '16' },
npm WARN EBADENGINE   current: { node: 'v17.9.0', npm: '8.5.5' }
npm WARN EBADENGINE }
```

That specific package.json said it required _precisely_ Node 16, which got interpreted as `16.0.0`.  I've tweaked it to accept `>= 16.0`, which should remove the warning; I highly doubt our package requires that precise of a version setting.

Finally, Developer's 'server' component had builds and tests falling over until ts-node was updated and the implicit "dom" library reference in its tsconfig was dropped, so `ts-node` got a global update, too.  (We had a few different versions floating around.)

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
While the comment linked in the description says we should update to Node.js 18, the only reason I've been able to infer so far is that it's the oldest version still actively maintained; 16 LTS is now too old.

If there _are_ other reasons, it'd be nice to document those... and to update any relevant in-repo package.json engine-requirement entries.

**Comment by jahorton** [comment]
Developer's build fell over after the KMW build component as it transitioned to developer/server:

```
[09:09:56 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5223&logView=flowAware&linesState=5223)
  /c/BuildAgent/work/7ac43416c45637e9/keyman/developer/src/server /c/BuildAgent/work/7ac43416c45637e9/keyman/developer/src/server

[09:09:57 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5224&logView=flowAware&linesState=5224)
  

[09:09:57 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5225&logView=flowAware&linesState=5225)
  > build

[09:09:57 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5226&logView=flowAware&linesState=5226)
  > tsc -b

[09:09:57 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5227&logView=flowAware&linesState=5227)
  

[09:10:03 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5228&logView=flowAware&linesState=5228)
  node_modules/@types/node/globals.d.ts(72,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; abort(reason?: any): AbortSignal; timeout(milliseconds: number): AbortSignal; }', but here has type '{ new (): AbortSignal; prototype: AbortSignal; }'.

[09:10:03 ](https://build.palaso.org/buildConfiguration/Keyman_Developer_Test/364140?buildTab=log&focusLine=5229&logView=flowAware&linesState=5229)
  Compilation failed.
```

Currently tracking that down; got a basic fix set, but then something else fell over that I'm still trying to resolve.

**Comment by jahorton** [comment]
Regarding the change on the most recent commit: https://www.typescriptlang.org/tsconfig#lib

> You may want to change these for a few reasons:
>
> *   Your program doesn’t run in a browser, so you don’t want the "dom" type definitions

The "dom" type definitions provided a different type spec for that in `@types/node`, which was the source of the conflict.  As `developer/server` compiles down to Node modules (commonjs), it's meant for command-line and not DOM, so dropping those type definitions should be safe.

**Comment by jahorton** [comment]
OK... this one's arising later, when trying to run the unit-test:

image

> Error: Debug Failure. False expression: Non-string value passed to `ts.resolveTypeReferenceDirective`, likely by a wrapping package working with an outdated `resolveTypeReferenceDirectives` signature.  This is probably not a problem in TS itself.

Hmm.  https://stackoverflow.com/a/72528793

> Upgrading to "ts-node": "10.8.1" solved the issue for me.

Well... we were already there, so this may take some investigation.

**Comment by jahorton** [comment]
OK, that really was about it, just with one wrinkle - I had to update the repo root's version of `ts-node`, not just the one referenced by `developer/server`.  Took me a bit to narrow down on the reason...

image

Updating `mocha` for just developer-server didn't resolve it; I'm guessing because the repo-root had its own TypeScript version, but it's hard to be sure.

**Comment by mcdurdin** [comment]
> I highly doubt our package requires that precise of a version setting.

What's wrong with a little bit of precision every now and then? 🤣

File: developer/src/server/tsconfig.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
     "noImplicitAny": true,
     "strictBindCallApply": true,
     "strictFunctionTypes": true,
-    "noUnusedLocals": true
+    "noUnusedLocals": true,
+    "lib": ["es6"]
```
Comment (mcdurdin) [inline developer/src/server/tsconfig.json:19]: > The "dom" type definitions provided a different type spec for that in `@types/node`, which was the source of the conflict. As `developer/server` compiles down to Node modules (commonjs), it's meant for command-line and not DOM, so dropping those type definitions should be safe.

Sounds good to me. Given we are now node18, can we target a newer version than ES6?

File: resources/build/version/package.json [hunk 2]
```diff
@@ -2,18 +2,18 @@
...
   "devDependencies": {
     "@types/node": "^13.7.0",
     "@types/semver": "^7.1.0",
     "semver": "^7.1.2",
-    "ts-node": "^8.6.2"
+    "ts-node": "^10.9.1"
   },
   "engines": {
-    "node": "16"
+    "node": ">=16.0"
```
Comment (mcdurdin) [inline resources/build/version/package.json:16]: nicer

**Review by mcdurdin — APPROVED** [review]
We should probably verify `lib` in tsconfig.json for our other node-based modules (kmlmc etc). This maybe should be done in the `feature-ldml` branch too.

This is independent of node version bump, yeah? Have we done that yet; I've lost track of this? (Important: we bundle node.exe with Developer so need to update that at the same time.)

File: developer/src/server/tsconfig.json (hunk 1 above)
Comment (jahorton) [reply developer/src/server/tsconfig.json:19]: Oh, almost certainly so.  I just went with something I knew was safe there.

**Comment by jahorton** [comment]
> We should probably verify `lib` in tsconfig.json for our other node-based modules (kmlmc etc). This maybe should be done in the `feature-ldml` branch too.
> 
> This is independent of node version bump, yeah? Have we done that yet; I've lost track of this? (Important: we bundle node.exe with Developer so need to update that at the same time.)

I haven't done anything toward this; where should I look to update what version Developer bundles?  I was operating under the notion that it only mattered when setting up build agents, hence [this comment from above](https://github.com/keymanapp/keyman/pull/8196#issuecomment-1423498631):

> While the comment linked in the description says we should update to Node.js 18 (LTS), the only reason I've been able to infer so far is that it's the oldest version still actively maintained; 16 LTS is now too old.
> 
> If there _are_ other reasons, it'd be nice to document those... and to update any relevant in-repo package.json engine-requirement entries.

Feel free to make any followup commits you deem necessary, should you find that easier.

Also, are there any dev-machine setup scripts to update?  I did check `resources/devbox`, but only found a macOS setup script - and that one doesn't seem to be version-locked.

File: developer/src/server/tsconfig.json (hunk 1 above)
Comment (jahorton) [reply developer/src/server/tsconfig.json:19]: After a little hunting, I came upon this link:  https://kangax.github.io/compat-table/es2016plus/

It appears to indicate we could say ES2022.

**Comment by mcdurdin** [comment]
> I haven't done anything toward this; where should I look to update what version Developer bundles? I was operating under the notion that it only mattered when setting up build agents, hence [this comment from above](https://github.com/keymanapp/keyman/pull/8196#issuecomment-1423498631):

`developer/src/inst/node/dist`

> > While the comment linked in the description says we should update to Node.js 18 (LTS), the only reason I've been able to infer so far is that it's the oldest version still actively maintained; 16 LTS is now too old.
> > If there _are_ other reasons, it'd be nice to document those... and to update any relevant in-repo package.json engine-requirement entries.

When updating a dependency, we'd normally update to the current stable supported version, yeah? I'm not sure what other reason you are looking for?

> Also, are there any dev-machine setup scripts to update? I did check `resources/devbox`, but only found a macOS setup script - and that one doesn't seem to be version-locked.

No, I don't think we have any other setup scripts, and that one is still only partly maintained.

**Comment by jahorton** [comment]
No instructions there on how to update - just a bundled .exe file.  I don't see any corresponding artifact at https://nodejs.org/en/download/, so I'm not sure how to proceed on this point.

Also...

https://github.com/keymanapp/keyman/blob/670932f92257d903c98d30427f87395185f486ad/developer/src/inst/node/dist/Keyman_README.md?plain=1#L3-L7

I'm guessing those are related to the `*.node` file entries in the subfolders there?  A readme in that folder seems to indicate I can run https://github.com/keymanapp/keyman/blob/master/developer/src/server/build-addins.inc.sh for those, at least.

**Comment by mcdurdin** [comment]
I'll take a look at updating node in Developer once I recover. Can you pop it into an new issue?

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.50-alpha](https://keyman.com/downloads/releases/alpha/17.0.50)

**Pull request merged**